[#25] 공연장 정보 추가 기능 구현

hanwix2 opened the pull request:
### Info
- 공연장(Venue) 정보(공연장 이름, 주소, 전화번호, 홈페이지)
 - DB: venue 테이블의 name 칼럼, index 추가

---

### To Do
- [x] 공연장 정보 추가 기능 구현
- [x] 공연장 이름이 중복된 경우 검사
- [x] 공연장 정보 관련 기능 수행 전 로그인 & 관리자 권한 검사
- [x] 공연장 정보 추가 로직에 트랜잭션 적용
- [x] 관련 테스트 코드

---

### ETC
- REDIS 적용 브랜치(feature/18/sh) merge: Redis 적용 후 (*커밋 99595fb*) 발생한 오류 해결 코드 적용

tlsgkr7416 committed:
Redis 사용시 serializer 되는 객체 처리

userSession에 Serializable 구현
Redis 사용 테스트 코드 수정

hanwix2 committed:
[#25] 공연장 정보 추가 기능 구현

 - 공연장(Venue) 정보(공연장 이름, 주소, 전화번호, 홈페이지)
 - 공연장 추가 요청 객체: VenueReqeuset
 - 공연장 관련 Controller: VenueController
 - 공연장 관련 Service: VenueService
 - 공연장 관련 Mapper: VenueMapper

hanwix2 committed:
[#25] 중복된 이름의 공연장 등록 요청에 대한 예외 처리

- 공연장 신규 등록 시 VenueService에서 동일한 공연장명 확인을 위한 메소드(checkVenueExists) 생성 및 기타 관련 코드 추가
- 동일한 공연장명이 존재할 시 예외(VenueAlreadyExistException)를 던지고 VenueExceptionAdvice에서 처리
- VenueRequest 클래스명 Venue 로 변경: 범용적으로 사용하기 위함

f-lab-dev commented:
여기 체크리스트를 보면 추가, 수정, 삭제가 다 있는데 같은 이슈에서 모든걸 다 처리하실 예정이신걸까요?

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -10,3 +10,13 @@ CREATE TABLE `user` (
+    `venueId`   INT UNSIGNED    NOT NULL    AUTO_INCREMENT,
+    `name`      VARCHAR(45)     NOT NULL,
+    `address`   VARCHAR(100)    NOT NULL,
+    `telNum`    VARCHAR(20)     NOT NULL,
```
전화번호를 쓰신 것 같은데 `tel`만 쓰셔도 의미를 전달하기에 충분하긴 합니다~

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -10,3 +10,13 @@ CREATE TABLE `user` (
+
+CREATE TABLE venue
+(
+    `venueId`   INT UNSIGNED    NOT NULL    AUTO_INCREMENT,
```
여기도 이름을 그냥 id로 지어도 `venue`테이블의 id니까 의미가 충분히 전달될 것 같네요~

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -0,0 +1,22 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/venue")
```
여기는 URL 원칙을 잘 지키고 있나요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueService.java:
```diff
@@ -0,0 +1,25 @@
+    private final VenueMapper venueMapper;
+
+    public void insertVenue(Venue venue) {
+        checkVenueExists(venue.getName());
+        venueMapper.insertVenue(venue);
```
이제 `@Transactional` 어노테이션에 대해서 알아보시면 좋을 것 같은데요~
이걸 안달아줬을 경우 이 구문에서 어떤 문제가 발생할 수 있는지 알아보시면 좋을 것 같아요

hanwix2 replied:
동시에 여러 관리자가 우연히 같은 이름의 공연장명으로 공연장 정보를 추가하려고 할 때 
충돌이 발생할 수 있을 것 같습니다.

여기서 충돌이란, 같은 공연장이 없다고 체크했지만, 그 사이 다른 곳에서 동일한 공연장을 등록하였고 DB에 중복되는 데이터가 존재할 수 있다는 것 입니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/advice/VenueExceptionAdvice.java:
```diff
@@ -0,0 +1,23 @@
+@RestControllerAdvice
+public class VenueExceptionAdvice {
+
+    @ExceptionHandler(VenueAlreadyExistsException.class)
```
👍

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -10,3 +10,13 @@ CREATE TABLE `user` (
+CREATE TABLE venue
+(
+    `venueId`   INT UNSIGNED    NOT NULL    AUTO_INCREMENT,
+    `name`      VARCHAR(45)     NOT NULL,
```
name이 검색조건으로 이용되던데 인덱스를 걸어주시면 좋을 것 같네요~

tlsgkr7416 committed:
테스트 코드 session위치 수정

테스트의 결합력을 낮추기 위해 session 위치 수정

hanwix2 committed:
[#25] 공연장 정보 추가 기능 관련 피드백 반영

- 공연장(Venue) 인스턴스명, 테이블 칼럼명 수정
- venue table index 추가(칼럼: name)
- VenueController url 수정: '/venue' -> 'venues'

hanwix2 committed:
Merge branch 'feature/18/sh' of https://github.com/f-lab-edu/show-ticketing-service into feature/18/sh

hanwix2 committed:
[#25] 공연장 정보 추가 요청 전 관리자 권한 검사

- 인터셉터(ManagerAuthorityCheckInterceptor)를 통해서 세션에 저장된 UserSession의 UserType으로 관리자 권한 검사
- @ManagerAuthority 어노테이션을 통해 권한 검사할 메소드 설정
- 관리자 권한이 없을 시 UserHaveNoAuthorityException 예외

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -15,6 +20,8 @@
     private final VenueService venueService;
 
     @PostMapping
+    @ManagerAuthority
```
이거는 기존에 있던 일반유저의 권한체크 어노테이션에 role만 추가해주면 이렇게 새로 만들지 않고 기존에 있던 것을 재활용하는게 가능하지 않을까요?

hanwix2 replied:
그게 좋을 것 같네요. 적용해보겠습니다!

hanwix2 committed:
[#25] 공연장 정보 추가 기능 테스트 코드 작성

hanwix2 committed:
[#25] 사용자 권한 검사 인터셉터를 기존의 로그인 검사 인터셉터로 병합

- LoginCheckInterceptor에서 로그인 검사와 동시에 세션에 저장된 유저의 권한 검사 진행
- ManagerAuthorityCheckInterceptor 삭제
- 로그인 검사에 사용하는 @UserAuthenticationNecessary 어노테이션에 권한을 의미하는 role 속성(AccessRole Enum / default: ALL_USER) 추가
- AccessRole Enum의 속성: GENERAL(일반 유저), MANAGER(관리자), ALL_USER(모든 사용자)

hanwix2 committed:
[#25] 공연장 등록 로직 트랜잭션 적용

- VenueService의 insertVenue 메소드에 적용
- DB의 default 격리수준을 따름

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/model/venue/Venue.java:
```diff
@@ -1,13 +1,15 @@
 import javax.validation.constraints.NotNull;
 
 @Getter
+@Builder
```
빌더 사용하신 곳이 없는 거 같은데 혹시 사용하신 이유가 따로 있으세요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/interceptor/LoginCheckInterceptor.java:
```diff
@@ -22,8 +28,17 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
+
+            UserSession userSession = (UserSession) request.getSession().getAttribute(USER);
+
+            if ((userAuthentication.role() == AccessRoles.GENERAL && userSession.getUserType() != UserType.GENERAL) ||
+                    (userAuthentication.role() == AccessRoles.MANAGER && userSession.getUserType() != UserType.MANAGER))
```
이 조건문이 꽤나 복잡한데 메소드로 분리하면 어떨까요? 그럼 if문이 더 읽기 좋아질 것 같습니다~

hanwix2 replied:
중첩된 if 문이 가독성이 떨어진다는 말씀이시죠? 
적으면서도 고민하던 부분이긴 합니다. 분리하겠습니다!

hanwix2 committed:
[#25] 로그인, 권한 검사 인터셉터 관련 피드백 적용

- LoginCheckInterceptor preHandler의 로그인 검사와 권한 검사에 사용되는 중첩된 if 문을 메소드로 분리

hanwix2 committed:
[#26] 공연장 정보 수정 기능 구현

- VenueController - PUT 메소드에 PathVariable로 공연장 id와 RequestBody로 Venue 객체를 전달
- VenueService, VenueMapper - 공연장 정보 수정 기능 로직인 updateVenueInfo 메소드 추가

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/interceptor/LoginCheckInterceptor.java:
```diff
@@ -29,18 +29,24 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
+            throw new UserNotLoggedInException();
+    }
+
+    private void roleCheck(UserAuthenticationNecessary userAuthentication, UserSession userSession) {
```
`roleCheck`라는 이름보다는 더 자세한 이름을 지어줄 수 있을 것 같아요~

hanwix2 committed:
[#25] LoginCheckInterceptor의 권한 검사 메소드명 변경

- roleCheck -> checkUserTypeMatchRole

hanwix2 committed:
Merge pull request #42 from f-lab-edu/feature/26/jh

[#26] 공연장 정보 수정 기능 구현

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/interceptor/LoginCheckInterceptor.java:
```diff
@@ -44,7 +44,7 @@ private void loginCheck() {
     }
 
-    private void roleCheck(UserAuthenticationNecessary userAuthentication, UserSession userSession) {
+    private void checkUserTypeMatchRole(UserAuthenticationNecessary userAuthentication, UserSession userSession) {
```
보통 이런 유효성검사에는 `validate`라는 단어를 많이 씁니다~

f-lab-dev reviewed (approved):
이름 부분만 잘 고민해보시고 머지하셔도 될 것 같네요~

hanwix2 committed:
[#25] LoginCheckInterceptor의 권한 검사 메소드명 변경

- checkUserTypeMatchRole() -> validateUserRole()
